storey/sources.py

```diff
@@ -703,17 +718,28 @@ async def _run_loop(self):
                     if isinstance(self._key_field, list):
                         key = []
                         for key_field in self._key_field:
+                            if body[key_field] is None:
+                                create_event = False
+                                break
                             key.append(body[key_field])
                     else:
                         key = body[self._key_field]
+                        if key is None:
+                            create_event = False
                 time = None
                 if self._time_field:
                     time = body[self._time_field]
                 id = None
                 if self._id_field:
                     id = body[self._id_field]
-                event = Event(body, key=key, time=time, id=id)
-                await self._do_downstream(event)
+                if create_event:
+                    event = Event(body, key=key, time=time, id=id)
+                    await self._do_downstream(event)
```

Comment:
Better move the work for creating the event into the if, to avoid unnecessary work.

Reply:
it;s in the if :)

Reply:
I meant the
```
                time = None
                if self._time_field:
                    time = body[self._time_field]
                id = None
                if self._id_field:
                    id = body[self._id_field]
```
stuff.